[StructuralMechanics] Clean property access in shells

philbucher opened the pull request:
This PR improves how the shells handle the access to the properties
So far they saved some things internally rather than accessing it each time it is needed.

This is a problem when @KratosMultiphysics/optimization want to compute the sensitivities, since they only exchange the properties and then reevaluate the stiffness matrix and other things.

Plus it requires more memory & internal management for updating the saved quantities.

Also this is a preparation for #2239 

I tried to minimize the changes, the most important ones are in `shell_cross_section.h/.cpp`, please have a close look in case I have overseen sth

Tests are running ofc

I will not do more cleanup in this PR, I know there are more things that can be done better.
I will do them in a follow-up PR

@MFusseder @armingeiser please test this with your new workflow

philbucher committed:
Moving Initialize to baseclass

philbucher committed:
minor in base-shell

philbucher committed:
Removing saving of thickness

philbucher committed:
Adding SHELL_OFFSET variable

philbucher committed:
Moved Orientationangle, Density and ShellOffset

philbucher committed:
Updateing Integration Points on the Fly

philbucher committed:
adding many const to avoid unwanted changes

philbucher committed:
minor

philbucher committed:
Compute Location on the fly

philbucher committed:
Cleanup

loumalouomega commented on applications/StructuralMechanicsApplication/custom_python/add_cross_sections_to_python.cpp:
```diff
@@ -62,7 +62,7 @@ void AddCrossSectionsToPython(pybind11::module& m)
     .def("AddPly", &ShellCrossSection::AddPly)
     .def("EndStack", &ShellCrossSection::EndStack)
-    .def("SetOffset", &ShellCrossSection::SetOffset)
+    // .def("SetOffset", &ShellCrossSection::SetOffset)
```
Remove instead of commenting?

philbucher replied:
I would like to remove this file altogether, requires a separate discussion beforehand though

loumalouomega commented on applications/StructuralMechanicsApplication/custom_utilities/shell_utilities.cpp:
```diff
@@ -16,139 +16,217 @@
+
+void InterpToStandardGaussPoints(double& rV1, double& rV2,double& rV3)
+{
+    double vg1 = rV1;
```
const?

loumalouomega reviewed (dismissed):
Minor comments

philbucher committed:
removing unused constructor arg

philbucher committed:
adding const

msandre commented on applications/StructuralMechanicsApplication/custom_utilities/shell_cross_section.cpp:
```diff
@@ -758,155 +714,95 @@ int ShellCrossSection::Check(const Properties& rMaterialProperties,
 
-        if(this->mStack.size() < 1)
-            KRATOS_THROW_ERROR(std::logic_error, "The Ply Stack of a ShellCrossSection cannot be empty", "")
+    if(this->mStack.size() < 1)
```
These could also be `KRATOS_ERROR_IF`.

msandre commented on applications/StructuralMechanicsApplication/custom_utilities/shell_cross_section.cpp:
```diff
@@ -758,155 +714,95 @@ int ShellCrossSection::Check(const Properties& rMaterialProperties,
-                    Ply::IntegrationPointCollection::size_type numip( iPly.NumberOfIntegrationPoints() );
-                    if(numip < 1)
-                        KRATOS_THROW_ERROR(std::logic_error, "The number of integration points in a Ply is not set properly", numip);
+        if(r_ply.GetPropertiesPointer() == NULL)
```
I think `NULL` is only kept for backwards compatibility in c++11. Why not use `nullptr` or `!' here?

RiccardoRossi replied:
+1 for nulltptr

philbucher replied:
This was not added by me

I would have wanted to cahnge all the error macros in a separate PR to not pollute this one but I can also do it here then

msandre replied:
I understand not wanting to expand this PR, however I commented this because it appears to be added. Still I think the NULL and KRATOS_THROW_ERROR are doable in this PR, but I am not insisting on it.

msandre commented on applications/StructuralMechanicsApplication/custom_utilities/shell_cross_section.cpp:
```diff
@@ -758,155 +714,95 @@ int ShellCrossSection::Check(const Properties& rMaterialProperties,
+
+            const ConstitutiveLaw::Pointer& iPointLaw = r_int_point.GetConstitutiveLaw();
+
+            if(iPointLaw == NULL)
```
And here.

msandre commented on applications/StructuralMechanicsApplication/custom_utilities/shell_cross_section.hpp:
```diff
@@ -414,36 +415,34 @@ class KRATOS_API(STRUCTURAL_MECHANICS_APPLICATION) ShellCrossSection : public Fl
-            : mThickness(0.0)
-            , mLocation(0.0)
-            , mOrientationAngle(0.0)
+            : mPlyIndex(0)
```
You removed `mOrientationAngle(0.0)` when you replaced `mThickness` and `mLocation` by `mPlyIndex`. I suspect this is accidental since now the member variable mOrientationAngle is not initialized in the default constructor. This style of default initialization is a known source of bugs which is one reason why I think there is a strong consensus that one should prefer in class member initialization for defaults (http://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Rc-default). I would suggestion always switching to in class initialization whenever you have to update a constructor that uses this buggish style.

philbucher replied:
I forgot to remove mOrientationAngle
Thx for spotting!

msandre commented on applications/StructuralMechanicsApplication/custom_utilities/shell_cross_section.hpp:
```diff
@@ -414,36 +415,34 @@ class KRATOS_API(STRUCTURAL_MECHANICS_APPLICATION) ShellCrossSection : public Fl
-        Ply(double thickness, double location, double orientationAngle, int numPoints, const Properties::Pointer & pProperties)
-            : mThickness(thickness)
-            , mLocation(location)
+        Ply(const int PlyIndex, int NumIntegrationPoints, const Properties::Pointer & pProperties)
```
Is it intentional to not initialized `mOrientationAngle`. Did you mean to completed remove this variable from the class?

philbucher replied:
Yep I forgot to completely remove it

msandre commented on applications/StructuralMechanicsApplication/custom_utilities/shell_cross_section.hpp:
```diff
@@ -1158,16 +1135,10 @@ class KRATOS_API(STRUCTURAL_MECHANICS_APPLICATION) ShellCrossSection : public Fl
-    		{
-    			mPlyConstitutiveMatrices[ply].resize(6, 6, false);
-    		}
+    		if (mBehavior == Thick) mPlyConstitutiveMatrices[ply].resize(8, 8, false);
```
Minor formatting issue. I don't think this inline is the normal way to format if statements.

philbucher replied:
I like it more this way :)
Can change it though if it is not the standard

msandre replied:
ok, it's minor so just leave it.

msandre reviewed (changes requested):
My main concern is about the orientation angle. Did you mean to completely remove it from the class?

RiccardoRossi commented on applications/StructuralMechanicsApplication/custom_utilities/shell_cross_section.cpp:
```diff
@@ -49,79 +47,62 @@ void ShellCrossSection::BeginStack()
-        }
-    }
-    ss << std::endl;
+    // ss << std::fixed;
```
why commenting this?

philbucher replied:
Leftover from debugging, will change it

armingeiser replied:
@philbucher ping

philbucher replied:
I am traveling today, will finish this PR tmr

RiccardoRossi commented on applications/StructuralMechanicsApplication/custom_utilities/shell_cross_section.cpp:
```diff
@@ -49,79 +47,62 @@ void ShellCrossSection::BeginStack()
 }
 
-void ShellCrossSection::AddPly(double thickness, double orientationAngle, int numPoints, const Properties::Pointer & pProperties)
+void ShellCrossSection::AddPly(const IndexType PlyIndex, int numPoints, const Properties::Pointer & pProperties)
```
why did u remove the thickness? it was a fundamental feature i would say...

philbucher replied:
It is NOT removed
It is now accesses through the Properties, the way it should be accesses

RiccardoRossi commented on applications/StructuralMechanicsApplication/custom_utilities/shell_cross_section.hpp:
```diff
@@ -594,18 +592,14 @@ class KRATOS_API(STRUCTURAL_MECHANICS_APPLICATION) ShellCrossSection : public Fl
-            rSerializer.load("T", mThickness);
-            rSerializer.load("L", mLocation);
-            rSerializer.load("O", mOrientationAngle);
+            rSerializer.load("idx", mPlyIndex);
```
i don't understand why all of this should be removed...

philbucher replied:
I only remove storing of the member

It is now accesses through the properties

RiccardoRossi reviewed (commented):
i don't understand this change...

philbucher commented:
This PR does **NOT** remove functionality!!!

It only removes storing of some variables that should be accesses through the properties 

Maybe Armin can give a better explanation than me :)

philbucher committed:
Addressed reviewers comments

philbucher commented:
@msandre @RiccardoRossi I addressed your comments, thx for the review!

this will in any case NOT be merged before @armingeiser and @MFusseder confirm that it is working as intended

armingeiser commented:
Unfortunately it does not work yet with our workflow. 
The problem seems to be that the Sections/Plys/SectionParameters store a property pointer (and also pointers to the geometry) of the element.
So if one changes the property of the element, the section does still have the old one.
I have the feeling that everything that belongs to the element (properties, geometry...) should not be stored at the section, but passed in every call from within the element.

philbucher committed:
Merge branch 'master' into StructMech/shells-access-props-correctly

RiccardoRossi commented on applications/StructuralMechanicsApplication/custom_elements/base_shell_element.cpp:
```diff
@@ -439,9 +487,10 @@ void BaseShellElement::CheckProperties(const ProcessInfo& rCurrentProcessInfo)
+        // TODO is this needed???? => it is, the dummy is needed for "Check" => unify!
         ShellCrossSection::Pointer dummySection = ShellCrossSection::Pointer(new ShellCrossSection());
         dummySection->BeginStack();
-        dummySection->AddPly(props[THICKNESS], 0.0, 5, pGetProperties());
```
one question: what happens if the thickness changes? for a large deformation shell that would definitely be a problem (although our shells as of now are large rotation but not large deformation)

philbucher replied:
currently: nothing :)

RiccardoRossi commented on applications/StructuralMechanicsApplication/custom_utilities/shell_utilities.h:
```diff
@@ -20,52 +20,69 @@
+using IndexType = std::size_t;
+
+template<class TVec>
+inline void ShapeFunc(double Xi, double Eta, TVec& rN)
```
well...this one is really in the geometry (LocalCoordinates). I am not asking to change it, just to take it into account for the future

philbucher replied:
Thx for the hint @RiccardoRossi 
I will consider it in future iterations!

RiccardoRossi reviewed (dismissed):
this change looks fine to me.

however
1 - i cannot claim i did a really detailed review (sorry)
2 - please take a look at my 2 comments

philbucher committed:
Removing storing of Properties in Ply

armingeiser reviewed (commented):
@philbucher Thanks for your effort!
Now the shell sections are working also for the semi-analytic sensitivity analysis (in my branch).
We don't need to reset the sections after perturbation anymore (previously done with a custom function at the shell elements - can be removed now).
Overall the changes look good to me, I approve, but would others give some time to review.

philbucher committed:
cleanup

philbucher commented:
@armingeiser I did the final cleanup, can you reapprove pls?

I also ran some more tests, seems to be ok

philbucher committed:
minor
